server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
         return readCollection(reader, Sets::newHashSetWithExpectedSize, Collections.emptySet());
     }
 
+    /**
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}}.
+     * The returned set is immutable.
+     *
+     * @return the set of objects
+     * @throws IOException if an I/O exception occurs reading the set
+     */
+    public <T> Set<T> readImmutableSet(final Writeable.Reader<T> reader) throws IOException {
+        int count = readArraySize();
+        // special cases small arrays, just like in java.util.Set.of(...)
+        if (count == 0) {
+            return Set.of();
+        } else if (count == 1) {
+            return Set.of(reader.read(this));
+        } else if (count == 2) {
+            return Set.of(reader.read(this), reader.read(this));
+        }
+        Object[] entries = new Object[count];
+        for (int i = 0; i < count; i++) {
+            entries[i] = reader.read(this);
+        }
+        @SuppressWarnings("unchecked")
+        T[] typedEntries = (T[]) entries;
+        return Set.of(typedEntries);
```

Comment:
`Set.of` throws an exception on duplicates on contrary to `Set.copyOf`. I wonder if we should document it in the method Javadoc to make sure that we can use this method if the written data is a `Set`.

Reply:
Sure thing added a couple words on that.